File: wait_for.py
"""
wait_for.py — `desktop_wait_for` polling primitive.

Cold-start UI flake is the single biggest source of wasted iterations:
the executor calls ``desktop_launch`` and immediately tries to click
something that hasn't been drawn yet.  ``desktop_wait_for`` lets it ask
the system "wake me when X is true" instead, so a slow Word load or a
half-loaded browser tab doesn't burn three iterations of failed clicks.

Targets supported
-----------------
* ``window_title=<substring>``  — any visible window matches
* ``element_name=<name>``       — a11y element resolves
* ``text=<visible label>``       — OCR / a11y text match (no click)
* ``window_id=<id>``            — specific window handle exists

Polling cadence is fixed at 0.5 s (cheap calls) with a configurable
timeout (default 15 s).  Returns ``{found, target, elapsed, ...}`` so
the model can branch on success.
"""

from __future__ import annotations

import asyncio
import time
from typing import Any


async def wait_for(
    *,
    backend,
    window_title: str = "",
    element_name: str = "",
    text: str = "",
    window_id: str = "",
    timeout: float = 15.0,
    poll_interval: float = 0.5,
) -> dict[str, Any]:
    """
    Poll until any of the supplied targets is observable, or the timeout
    elapses.  At least one target MUST be supplied.

    Each target is checked in a forgiving way: errors during a poll
    (e.g. AT-SPI hiccups) are swallowed and the loop tries again.
    """
    targets = {
        "window_title": window_title,
        "element_name": element_name,
        "text": text,
        "window_id": window_id,
    }
    targets = {k: v for k, v in targets.items() if v}
    if not targets:
        return {
            "error": "wait_for requires at least one of "
                     "window_title, element_name, text, window_id.",
            "found": False,
        }

    # Normalize timeout once so the loop-exit comparison and the sleep
    # math both use the same effective ceiling.  A caller-supplied
    # timeout below the floor is silently raised so we still poll at
    # least once before giving up.
    timeout = max(0.5, float(timeout))
    poll_interval = max(0.1, float(poll_interval))
    start = time.monotonic()
    deadline = start + timeout

    last_observation: dict[str, Any] = {}

    while True:
        elapsed = time.monotonic() - start

        try:
            if window_title or window_id:
                listing = await backend.list_windows()
                wins = listing.get("windows") if isinstance(listing, dict) else listing
                wins = wins or []
                for w in wins:
                    title = str(w.get("title", "")) or ""
                    wid = str(w.get("id", "")) or ""
                    if window_title and window_title.lower() in title.lower():
                        return _ok(start, "window_title", w)
                    if window_id and window_id == wid:
                        return _ok(start, "window_id", w)
                last_observation["windows"] = len(wins)
        except Exception as e:
            last_observation["window_error"] = str(e)

        if element_name:
            try:
                if hasattr(backend, "find_element"):
                    el = await backend.find_element(name=element_name)
                    if el and not el.get("error") and el.get("rect"):
                        return _ok(start, "element_name", el)
            except Exception as e:
                last_observation["element_error"] = str(e)

        if text:
            try:
                if hasattr(backend, "find_text_on_screen"):
                    found = await backend.find_text_on_screen(text, 0)
                    if found and found.get("found"):
                        return _ok(start, "text", found)
            except Exception as e:
                last_observation["text_error"] = str(e)

        if elapsed >= timeout:
            return {
                "found": False,
                "target": None,
                "elapsed": round(elapsed, 3),
                "timeout": timeout,
                "targets": targets,
                "last_observation": last_observation,
            }
        # Sleep until next poll, but never past the deadline.
        await asyncio.sleep(min(poll_interval, max(0.05, deadline - time.monotonic())))


def _ok(start: float, kind: str, observation: Any) -> dict[str, Any]:
    return {
        "found": True,
        "target": kind,
        "elapsed": round(time.monotonic() - start, 3),
        "observation": observation,
    }

File: pi-extension/src/wait_for.ts
/**
 * wait_for.ts — `desktop_wait_for` polling primitive.
 *
 * Mirror of Python ``wait_for.py``.  Polls the backend until any of
 * window_title / element_name / text / window_id resolves, or the
 * timeout elapses.
 */

import type { DesktopBackend } from "./types.js";

export interface WaitForResult {
  found: boolean;
  target?: string;
  elapsed: number;
  timeout?: number;
  observation?: unknown;
  targets?: Record<string, string>;
  lastObservation?: Record<string, unknown>;
  error?: string;
}

export interface WaitForOptions {
  windowTitle?: string;
  elementName?: string;
  text?: string;
  windowId?: string;
  timeout?: number;
  pollInterval?: number;
}

export async function waitFor(
  backend: DesktopBackend,
  options: WaitForOptions,
  signal?: AbortSignal,
): Promise<WaitForResult> {
  const targets: Record<string, string> = {};
  if (options.windowTitle) targets.window_title = options.windowTitle;
  if (options.elementName) targets.element_name = options.elementName;
  if (options.text) targets.text = options.text;
  if (options.windowId) targets.window_id = options.windowId;

  if (Object.keys(targets).length === 0) {
    return {
      found: false,
      elapsed: 0,
      error: "wait_for requires at least one of window_title, element_name, text, window_id.",
    };
  }

  const timeout = Math.max(0.5, options.timeout ?? 15.0);
  const pollInterval = Math.max(0.1, options.pollInterval ?? 0.5);
  const start = Date.now();
  const deadline = start + timeout * 1000;
  const lastObservation: Record<string, unknown> = {};

  while (true) {
    const elapsed = (Date.now() - start) / 1000;

    if (options.windowTitle || options.windowId) {
      try {
        const r = await backend.listWindows(signal);
        const wins = r.windows ?? [];
        for (const w of wins) {
          const title = String(w.title ?? "");
          const id = String(w.id ?? "");
          if (options.windowTitle && title.toLowerCase().includes(options.windowTitle.toLowerCase())) {
            return { found: true, target: "window_title", elapsed: elapsedSec(start), observation: w };
          }
          if (options.windowId && options.windowId === id) {
            return { found: true, target: "window_id", elapsed: elapsedSec(start), observation: w };
          }
        }
        lastObservation.windows = wins.length;
      } catch (e) {
        lastObservation.window_error = (e as Error).message;
      }
    }

    if (options.elementName && backend.findElement) {
      try {
        const el = await backend.findElement({ name: options.elementName }, signal);
        if (el && el.rect) {
          return { found: true, target: "element_name", elapsed: elapsedSec(start), observation: el };
        }
      } catch (e) {
        lastObservation.element_error = (e as Error).message;
      }
    }

    if (elapsed >= timeout) {
      return {
        found: false,
        elapsed: elapsedSec(start),
        timeout,
        targets,
        lastObservation,
      };
    }

    const remaining = (deadline - Date.now()) / 1000;
    const sleep = Math.max(0.05, Math.min(pollInterval, remaining));
    await new Promise((r) => setTimeout(r, sleep * 1000));
    if (signal?.aborted) {
      return { found: false, elapsed: elapsedSec(start), error: "aborted" };
    }
  }
}

function elapsedSec(start: number): number {
  return Math.round(((Date.now() - start) / 1000) * 1000) / 1000;
}
